src/librustc_typeck/check/mod.rs

```diff
@@ -3687,6 +3687,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// If `expr` is a `match` expression that has only one non-`!` arm, use that arm's tail
+    /// expression's `Span`, otherwise return `expr.span`. This is done to give bettern errors
+    /// when given code like the following:
+    /// ```text
+    /// if false { return 0i32; } else { 1u32 }
+    /// //                               ^^^^ point at this instead of the whole `if` expression
+    /// ```
+    fn get_expr_coercion_span(&self, expr: &hir::Expr) -> syntax_pos::Span {
+        if let hir::ExprKind::Match(_, arms, _) = &expr.node {
+            let arm_spans: Vec<Span> = arms.iter().filter_map(|arm| {
+                self.in_progress_tables
+                    .and_then(|tables| tables.borrow().node_type_opt(arm.body.hir_id))
+                    .and_then(|arm_ty| {
+                        if arm_ty.is_never() {
+                            None
+                        } else {
+                            Some(match &arm.body.node {
+                                // Point at the tail expression when possible.
+                                hir::ExprKind::Block(block, _) => block.expr
+                                    .as_ref()
+                                    .map(|e| e.span)
+                                    .unwrap_or(block.span),
+                                _ => arm.body.span,
+                            })
+                        }
+                    })
+            }).collect();
+            if arm_spans.len() == 1 {
```

Comment:
what happens for matches with more arms? Could we generalize this? Or is this a non-issue for matches?

Reply:
It's not how many arms the match has but rather how many are !. If all are ! But one, that one will be pointed at, otherwise younger the current behavior of pointing at the whole match expression.

Reply:
Right, but if there are more that are not !, why not also point to their trailing expressions?

Reply:
Because `ObligationCause` takes a `Span` and not a `MultiSpan` and it didn't occur to me to change it 😅

I'm not sure the output will be nice, but I will try it out locally at least and post the results.

Reply:
Thinking about it, it'll probably be too noisy. When the `spans.len() > 4` it'll be way more annoying than a single span pointing at the whole expression.

Reply:
Could we include tests for `match`?

Reply:
@Centril done.

Reply:
@oli-obk the amount of churn to get `ObligationCause` work nicely with a `MultiSpan` is much higher than I wanted for this PR :-|

Is it ok if we leave that experiment for a follow up PR?

Reply:
yes